### PR TITLE
Removed allocations for transpose/adjoint - diagonal multiplications

### PR DESCRIPTION
**Pull request by mkhrenov**
The dgmm! based implementation of multiplication of transpose/adjoint matrices with diagonal matrices that got merged makes a call to CuMatrix on the transpose/adjoint. This results in allocations and a substantial performance penalty. I haven't been able to figure out how to do this in place using the cuBLAS functions, but writing it out as in-place broadcast operations seems to work well. Open to any suggestions.

### CODE REVIEW

File: lib/cublas/linalg.jl
Comment (amontoison) [inline lib/cublas/linalg.jl]:
```suggestion
    C .*= transpose(B.diag)
```

Comment (amontoison) [reply lib/cublas/linalg.jl]: You don't want to conjugate the coefficients if `B` is a complex diagonal matrix.

File: lib/cublas/linalg.jl
Comment (amontoison) [inline lib/cublas/linalg.jl]: Why do you conjugate `B`?

**Review by amontoison — CHANGES_REQUESTED** [review]
I don't understand how you do some products, `M * D` and `D * M` gives different results and based of your broadcast relations, you can only do one them.
You use the trick of the adjoint / transpose like `D * M = (M' * D')'` but it means that you need to conjugate the result `C` at the end.

If you find a way to do that such that all tests passed, it's fine for me to use broadcast.

**Comment by amontoison** [comment]
I don't think that we can use `dgmm` without allocation for transpose/adjoint - diagonal multiplications because we can't reinterpret the input matrices as matrices stored in row-major format. So broadcast is the only alternative if we don't want allocations.

File: lib/cublas/linalg.jl
Comment (amontoison) [reply lib/cublas/linalg.jl]: Hum, I think you use a trick that I don't understand because you don't transpose `A`.
Can you explain what you do exactly?

Comment (mkhrenov) [reply lib/cublas/linalg.jl]: Sure, so the transpose (which I've put in both functions) of B.diag isn't a result of A being Transpose or Adjoint, it's just to make the broadcast multiplication work on the correct dimension for B diagonal. A being Adjoint or Transpose gets realized during the broadcasted assignment to C in all cases.

File: lib/cublas/linalg.jl
Comment (mkhrenov) [reply lib/cublas/linalg.jl]: It should have been a transpose, fixed.

**Review by amontoison — APPROVED** [review]

**Comment by mkhrenov** [comment]
Benchmarks seem to have been interrupted for some reason, but everything else looks good. Anything else I should do?

**Comment by amontoison** [comment]
LGTM but I can't merge the PR if all tests didn't passed.
You need Tim to manually merge the PR.

**Pull request merged**
